homeassistant/components/tplink/common.py

```diff
@@ -0,0 +1,202 @@
+
+
+ATTR_CONFIG = 'config'
+CONF_DISCOVERY = 'discovery'
+CONF_LIGHT = 'light'
+CONF_SWITCH = 'switch'
+
+
+class SmartDevices:
+    """Hold different kinds of devices."""
+
+    def __init__(
+            self,
+            lights: List[SmartDevice] = None,
+            switches: List[SmartDevice] = None
+    ):
+        """Constructor."""
+        self._lights = lights or []
+        self._switches = switches or []
+
+    @property
+    def lights(self):
+        """Get the lights."""
+        return self._lights
+
+    @property
+    def switches(self):
+        """Get the switches."""
+        return self._switches
+
+    def has_device_with_host(self, host):
+        """Check if a devices exists with a specific host."""
+        for device in self.lights + self.switches:
+            if device.host == host:
+                return True
+
+        return False
+
+
+async def async_get_discoverable_devices(hass):
+    """Return if there are devices that can be discovered."""
+    from pyHS100 import Discover
+
+    def discover():
+        devs = Discover.discover()
+        return devs
+    return await hass.async_add_executor_job(discover)
+
+
+async def async_discover_devices(
+        hass: HomeAssistantType,
+        exclude_devices: SmartDevices
+) -> SmartDevices:
+    """Get devices through discovery."""
+    _LOGGER.debug("Discovering devices")
+    devices = await async_get_discoverable_devices(hass)
+    _LOGGER.info(
+        "Discovered %s TP-Link smart home device(s)",
+        len(devices)
+    )
+
+    lights = []
+    switches = []
+
+    for dev in devices.values():
+        # If this device is configured statically, ignore dynamic setup.
+        if exclude_devices.has_device_with_host(dev.host):
+            continue
+
+        if isinstance(dev, SmartPlug):
+            try:
+                if dev.is_dimmable:  # Dimmers act as lights
+                    lights.append(dev)
+                else:
+                    switches.append(dev)
+            except SmartDeviceException as ex:
+                _LOGGER.error("Unable to connect to device %s: %s",
+                              dev.host, ex)
+
+        elif isinstance(dev, SmartBulb):
+            lights.append(dev)
+        else:
+            _LOGGER.error("Unknown smart device type: %s", type(dev))
+
+    return SmartDevices(lights, switches)
+
+
+def async_get_static_devices(config_data) -> SmartDevices:
+    """Get statically defined devices in the config."""
+    _LOGGER.debug("Getting static devices")
+    lights = []
+    switches = []
+
+    for type_ in [CONF_LIGHT, CONF_SWITCH]:
+        for entry in config_data[type_]:
+            host = entry['host']
+
+            if type_ == CONF_LIGHT:
+                lights.append(SmartBulb(host))
+            elif type_ == CONF_SWITCH:
+                switches.append(SmartPlug(host))
+
+    return SmartDevices(
+        lights,
+        switches
+    )
+
+
+def async_add_entities_retry(
```

Comment:
Should this be something that should be located in the core? I haven't reviewed the code, but I suppose this is/can be a common issue, and having such a feature in the core would save everyone's time and make the integrations simpler to develop. Just a thought :-)

Reply:
I think something like this should be in the core and actually wrote it with that intention. I figured I'd hold off submitting a PR for core until after this is scrutinized in the tplink component and we see this working in the wild.